### DISCUSSION TITLE
How is FPS computed?

### DISCUSSION THREAD

**Discussion by felix-knopp** [ultralytics/yolov5#6342] [Q&A] [answered]
How would I compute the FPS of my model in webcam detection?
I think the times listed at the end of val.py using batch-size 1 should give me the correct timings for real time detection.
However, when I start detect.py with webcam and average out the times that are listed after each detection I get way lower times per detection than the val.py results.

TL;DR: How to get time per detection for real time detection?

**Comment by glenn-jocher** [comment, marked as answer]
@felix-knopp up to you. In iDetection we display real-world inference FPS which maxes out at 30, which is the camera FPS. We also display inference time per frame (including NMS) in milliseconds:

<img src="https://user-images.githubusercontent.com/26833433/98019191-be36e000-1e01-11eb-97e0-69ca25c30531.jpg" width="800">

**Reply by felix-knopp** [reply]
@glenn-jocher Okay so 21.8 ms would result in about 45 FPS. Obviously drawing the boxes and displaying the result also takes time so that's probably why it's down to 30 in that case. 
I just saw that detect.py is also displaying inference time (not time per frame):
https://github.com/ultralytics/yolov5/blob/0cf932bf6346909189c53b375ef97551bb0c2326/detect.py#L178-L179

So I guess moving down `LOGGER.info(f'{s}Done. ({t3 - t2:.3f}s)') ` to the end of the detection loop and also measuring t3 there would be the most accurate way to get time per frame.

Edit: I also just noticed that t3 is taken before NMS in detect.py. So that's probably why my results differ from what i measured with val.py. 

Edit2: I added 
`if webcam:`
&emsp; `frame_time = time_sync() - t1`
&emsp; `total_time += frame_time`
&emsp; `total_predictions += 1`
&emsp; `print('FPS: %s\nAvg FPS: %s' % (1/frame_time, total_predictions/total_time))`
to the end of that loop:
https://github.com/ultralytics/yolov5/blob/0cf932bf6346909189c53b375ef97551bb0c2326/detect.py#L115
total_time and total_predictions is initialised with 0 before that loop

That does the trick:
![image](https://user-images.githubusercontent.com/74195458/150131769-90b2901d-5972-43ae-bcfb-399b77328bc8.png)

**Reply by glenn-jocher** [reply]
@felix-knopp no it's 30 FPS because the camera stream is delivered at 30 FPS. The FPS in this case will never exceed 30 no matter how fast the inference is. That's why we also provide the inference+NMS time separately, otherwise it would be saying 45 FPS on a 30 FPS stream which would raise some eyebrows.

**Reply by sainsaji** [reply]
how do i display this fps on the webcam frame ? I've done it on the console
![image](https://user-images.githubusercontent.com/95070425/169746655-0bdf1cd6-2f18-4b28-922d-034025e66403.png)

**Reply by sainsaji** [reply]
But the fps seems unbelievably high

**Reply by felix-knopp** [reply]
@sainsaji FPS is calculated by
(1s)/(computation time to make and show prediction)

In your case that last prediction took 0.020s which should result in 1s/0.020s = 50 FPS.

Somehow you got your timings wrong for your fps calculation. And for some reason it seams to be about 10x of the actual fps.

**Reply by sainsaji** [reply]
i'll divide by 10 and see.

**Reply by sainsaji** [reply]
> @sainsaji FPS is calculated by (1s)/(computation time to make and show prediction)
> 
> In your case that last prediction took 0.020s which should result in 1s/0.020s = 50 FPS.
> 
> Somehow you got your timings wrong for your fps calculation. And for some reason it seams to be about 10x of the actual fps.

made this small change and its working fine. The frame rate sometimes go zero throwing me out a division by zero exception.


while(frame_time!=0):
            print("FPS:")
            print(1/frame_time)
            break

![image](https://user-images.githubusercontent.com/95070425/169762615-195baaff-23df-44ec-87ff-e77390490b8a.png)

**Reply by sainsaji** [reply]
@felix-knopp  . Do you know any means to display it on the webcam frame as well?

**Reply by felix-knopp** [reply]
I highly doubt that your calculation is correct as it's always the EXACT same value which is really unlikely. 


> i'll divide by 10 and see.

Please don't just divide by 10. try to find your initial error. It might be coincidence that your wrong values are ~10x of the expected FPS

> @felix-knopp . Do you know any means to display it on the webcam frame as well?

I have not done this but you might be able to do that here:
https://github.com/ultralytics/yolov5/blob/cee5959c74336ea6a8b73e6ee99c6a5f45d64a4d/detect.py#L166-L172

Add something like `annotator.text((100, 100), FPS )` in line 172 outside of the loop.

Use your FPS from previous iteration. This should draw it 100px from top left corner.
syntax for the text function is here:

https://github.com/ultralytics/yolov5/blob/cee5959c74336ea6a8b73e6ee99c6a5f45d64a4d/utils/plots.py#L118-L121


I have not tried any of this myself. Try it for yourself.

**Reply by sainsaji** [reply]
That threw an error.
Traceback (most recent call last):
  File "detect.py", line 288, in <module>
    main(opt)
  File "detect.py", line 283, in main
    run(**vars(opt))
  File "D:\anaconda\envs\kunji\lib\site-packages\torch\autograd\grad_mode.py", line 26, in decorate_context
    return func(*args, **kwargs)
  File "detect.py", line 182, in run
    annotator.text((100, 100), 1/(t3-t2))
  File "C:\Users\Sain Saji Augustine\newproject\yolov5\utils\plots.py", line 120, in text
    w, h = self.font.getsize(text)  # text width, height
AttributeError: 'Annotator' object has no attribute 'font'
